datafusion/src/optimizer/simplify_expressions.rs

```diff
@@ -442,6 +442,8 @@ impl<'a> ExprRewriter for Simplifier<'a> {
         use Expr::*;
         use Operator::{And, Divide, Eq, Multiply, NotEq, Or};
 
+        let is_boolean_expr = self.is_boolean_type(&expr);
```

Comment:
I recommend putting this into the case statement to avoid evaluating it on each `Expr` (otherwise I think it ends up being `O(N^2)` as `is_boolean_type` recursively checks all inputs